### ISSUE TITLE
Rethinking parts of installation

### ISSUE THREAD

**Issue by archseer** [linux-surface/linux-surface#32]
Some ideas I had while tweaking the install guide.

https://github.com/linux-surface/linux-surface/wiki/Installation-and-Setup

Things I want to do:
- stop having setup.sh install kernel & wacom packages
- possibly have a ./download.sh <optional: arch/debian/fedora> that would download packages for you and echo the command needed to run
  - ok, if we do that, maybe have the setup script call that script?
- package the firmware we have as a package, similar to linux-firmware (instead of copying files into /lib directly)
- stop shipping the kernel with surface_sam and offer it as a dkms package, that way it's easier to upgrade separately (or for users to test, we don't need to compile a full kernel for them), and in another iteration or two it'll work
- add a pkg/ folder containing pkg/arch, pkg/fedora etc. for the kernel + acpi module. Those will contain up-to-date PKGBUILDS/build scripts that the user can use instead of pre-built packages.

The rationale is that with the upcoming simpler setup.sh (#22), the only distro-specific aspect is just installing/building the packages. This would make the installation guide a lot simpler, because it would have a lot less notices and conditionals (right now it's "if debian do this, but if arch, either use this other package you need to download manually or look at this third-party repo, or something else entirely, and say no in `setup.sh` 😅")

It would look something like:

[...] follow the guide to setup linux
1. run `./setup.sh`
2. obtain kernel & other packages
  a) run ./download.sh <distro> to fetch your packages 
  b) _or_ use the package repository
  c) _or_ build them yourself (use pkg/<distro>)
3. install all the things! (`pacman -U *.tar.zstd`)
4. update bootloader
5. reboot
6. post install
6.1 do general post install steps
6.2 do device specific post install steps by looking at the device's wiki page that addresses various quirks

**Comment by StollD** [comment]
> stop shipping the kernel with surface_sam and offer it as a dkms package, that way it's easier to upgrade separately (or for users to test, we don't need to compile a full kernel for them), and in another iteration or two it'll work

I am honestly not a fan of this. The reason is, that for any distro that ships with proper lockdown patches for secure boot (Fedora and I think Debian too), you will have to disable secure boot. With the lockdown patches and secure boot enabled the kernel will refuse to load any module that is not signed. You can sign DKMS modules yourself, but that adds another step of stuff the user has to do, which we wanted to remove by presigning the kernel.

Also, I am not a big fan of having another download script. The package repository is there, why not use it? Same goes for setup.sh: I think you can do everything it does right now using a package, and stop having it entirely. That way users would get automatic updates to the configs too, without having to run setup.sh again and again.

**Comment by archseer** [comment]
> You can sign DKMS modules yourself, but that adds another step of stuff the user has to do, which we wanted to remove by presigning the kernel.

Alright, but we could still package the acpi module as a separate package, and sign that one, right? I'd really like to keep the two separate. Besides the ACPI patches, the newer devices can work on an unpatched kernel. After 5.5, just one patch will be needed, which will land in 5.6.

I currently have secure boot disabled and was installing the ACPI module separately, which was great because it didn't take me an hour to rebuild every time more commits were pushed.

> Same goes for setup.sh: I think you can do everything it does right now using a package, and stop having it entirely. 

Yeah I kind of like the idea of getting rid of `setup.sh` completely. We could make a `surface-config` package, but I kind of wonder how that would work if it conflicted other package's files. With the cleanup that happened though, maybe we could just turn the configs into device documentation?

Currently we have:
- An udev rule for SP4
- Surface Go touchscreen systemd rule
- Options config for ath10k which is for Surface Go/SL3 (AMD), but also requires a fixed firmware from Killer which isn't included in the repo

None of these are universal, although these won't have any issues on devices that don't need the config.

- Options config for audio which fixes SB2 audio (but might mess with other platforms)
- NetworkManager stuff for the marvell chip

These most probably shouldn't be universally applied.

It's extra friction, but we could just have the user make all these configuration changes on his own, by following the device-specific guide. But there's already extra steps the user has to do on their own anyway (installing additional firmware, etc.)

> The package repository is there, why not use it? 

I think that's probably a safe default and we should use it in the guide, then only briefly mention the other choices (building it yourself or grabbing packages) for advanced users. At the moment there's a whole bunch of different ways mentioned so it can get confusing.

**Comment by StollD** [comment]
> Alright, but we could still package the acpi module as a separate package, and sign that one, right?

That could work, but it makes us way more work, because we have to sign two things: The kernel and the module. Additionally we would have to package it in such a way that the module build can reuse the tree that is used for the kernel, or we would have to make sure that the package is always built on a machine that is already running the latest kernel. Otherwise we risk that the ABI becomes incompatible and the module doesn't load anymore.

Out-of-tree modules are simply not a first class citizen with linux. DKMS makes packaging them easier, but it makes signing them way harder. I think rebuilding the kernel once a significant number of changes has been made is the better solution: it is not neccessarily *faster* but it is *simpler*. And we could look into automating it with GitHub Actions or sth. else.

@qzed What do you think?

> Yeah I kind of like the idea of getting rid of `setup.sh` completely. We could make a `surface-config` package, but I kind of wonder how that would work if it conflicted other package's files.

With the cleanup that is going on, there shouldn't be any conflicting files anymore. The only thing that might be risky are the modprobe files, but we could prefix them with `99-surface-` or move them all into a shared `99-surface.conf` file.

> These most probably shouldn't be universally applied.

They have been applied universally forever now, without any problems. We could look into scripting the network manager config (not neccessarily with udev, that didn't work well for me, but maybe with systemd: if `lsmod` contains `mwifiex`, disable power saving).

**Comment by qzed** [comment]
So just that I understand it correctly: With the lockdown patches, modules need to be signed to be loaded. To distribute signed modules, we'll need to distribute them in binary form. But modules can only be loaded on the kernel they are built with/for, right (IIRC there is a hard check for that and the kernel will refuse to load the module)? So it's pretty much impossible for us as we'd have to create a separate package for every distribution/kernel version which will break for every user as soon as the distribution decides to update the kernel, so the user needs to wait until we have the time to build another package. I'd argue that's pretty much impossible.

Is the private key distributed somewhere with the system? Looking at the [Gentoo](https://wiki.gentoo.org/wiki/Signed_kernel_module_support#Automatically_signing_kernel_modules_.28Portage.29) and [Arch](https://wiki.archlinux.org/index.php/Signed_kernel_modules#package-headers()) wikis it seems like (although not using the same file-name?). If so, then there's at least the option to add that to the packaging step.

Also I don't see any real reason for not including the SAM driver in the pre-built kernels. If you absolutely want to have the latest version of the driver you can always build it as module yourself, and the kernels should in any way be update fairly regularly. I can see why people would want to install only the module on 5.6 and later since some devices will then only need that and no additional patches (and we should continue to provide that option via the DKMS package), but many devices still need additional patches. And you can always blacklist all of the `surface_sam_...` modules and then just DKMS-install `surface_sam`. The module names are different so I think this should work (apart from some warnings of duplicated symbols maybe).

So I agree with @StollD that this would only complicate things. Also there are apparently distributions with no DKMS support at all (see e.g. https://github.com/qzed/linux-surfacebook2-mshw0153/issues/4), so we'll still have to provide patches.

On the other things:

- _stop having setup.sh install kernel & wacom packages:_ Makes sense, I'm not a big fan of scripts installing packages, I think we should let the user do that and refer to the provided repositories for that.
- _have a ./download.sh:_ Same as the previous one, we should just refer to the repositories. If anyone still wants to do that manually, sure why not. But I don't think there's a need to make that even more easier than following a list of links and installing the downloaded packages manually.

  E.g. 1) setup repositories, 2) run setup.sh, 3) `<your-aur-helper> -S a b c ...` on Arch.
- _package the firmware we have as a package:_ To be honest, I had my own non-AUR package for my SB2 for a long time now and I think it's time to provide something like that for all devices and distributions. I think that's the only thing we don't have consistent packaging for. IIRC for Arch there were packages on AUR, but I'm not sure how up-to-date they are. Last time I checked I didn't like the approach because the PKGBUILD was downloading and packaging stuff for _all_ devices and even including the linux-firmware repo. I think that should all be separate, so one package per device, only containing that specific IPTS firmware. 

- _add a pkg/ folder containing pkg/arch, pkg/fedora etc. for the kernel:_ This will work on Arch and Fedora, but probably be a pain for Debian. Also the more I think about packaging, the more I think we should separate those things into different repositories.

Some thoughts the config stuff: Getting rid of the `setup.sh` script completely would be neat, but I don't think that's possible. I think first of all we should clarify that the config files provided here are a starting point addressing issues across all devices (and that not everything is needed for every device) and encourage users to not run the setup script but instead help them to figure out what they need themselves via the wiki. But then again this can be quite a bit of work for the user, so we'll always need some sort of automated config install.

Providing a set of curated configs for each device would make sense, but then those configs not only depend on the device itself, but also some distribution/setup specific stuff (e.g. is NetworkManager installed?, is systemd installed?) so I'm not convinced that that's feasible. By providing a big one-stop config setup and some sort of knowledge-base via the wiki, I think we can help those who can't be bothered/have no time to look into that and at the same time motivate the ones who are interested to tweak things.

**Comment by archseer** [comment]
I agree that the repositories should probably be the official way to install everything. It would be as easy as 1) setup the repositories 2) install all the things via your package manager.

What I do like about github releases though, is that it's hosted on Github infrastructure instead of a 3rd party host. We're all able to add new releases currently, but if @StollD becomes inactive or in the unlikely case that tmsp.io lapses, that's not good. For example, phuntoo.org is mentioned on the wiki as the Gentoo repository, but doesn't resolve for me at all. We also need to keep the repo continually updated (latest package in there is 5.3).

The situation we're in right now is also a bit awkward because it's better for old devices to stay on 5.3, whereas the new devices need 5.4. With Arch, you usually can only install the latest package version.

> And you can always blacklist all of the surface_sam_... modules and then just DKMS-install surface_sam. 

True, I agree, I think I only separated the two because I had some issues with blacklisting `i915_legacy` back then, which isn't present anymore. I think we can keep the kernel as is, and I'll just offer a separate acpi PKGBUILD for those that want it.

> I didn't like the approach because the PKGBUILD was downloading and packaging stuff for all devices and even including the linux-firmware repo. I think that should all be separate, so one package per device, only containing that specific IPTS firmware.

I like the idea of a surface-firmware, but I'd like to include all the firmware we have (without linux-firmware). I don't think the user having to figure out what firmware packages they need is good UX, especially since it's just ~1-2MB per device. Same way `iwlwifi/firmware` covers all the intel wifi chips, or `ath10k-firmware` etc. Having one package means we will never get bug reports because someone installed the wrong package.

> add a pkg/ folder containing pkg/arch, pkg/fedora etc. for the kernel: This will work on Arch and Fedora, but probably be a pain for Debian. Also the more I think about packaging, the more I think we should separate those things into different repositories.

I'm okay with a separate repo, although I find one monorepo a lot easier to manage. I would like to have the build files that are used for the official packages available so that advanced users can reproduce the build without much extra hassle. The alternate configs were never merged into jakeday's repo so now they're all tracked separately and not always necessarily current. As a result, [a lot of issues are reported downstream](https://github.com/dmhacker/arch-linux-surface/issues) and they never reach us.

> By providing a big one-stop config setup and some sort of knowledge-base via the wiki, I think we can help those who can't be bothered/have no time to look into that and at the same time motivate the ones who are interested to tweak things.

I agree.

So I think we should offer:

- `linux-surface(-headers)` for the kernel, just like today
- `surface-firmware` as one package with all the device firmware
- `surface-config` that would contain all the config files under `root/` and then leave a notice in the install guide that the user might want to cherry pick the changes they want
- Unofficially, `surface-acpi/surface-sam(-dkms)` that wouldn't be mentioned in the install guide, for testers or those that want to run the latest changes
- Finally, drop `setup.sh`

I'm already maintaining my own PKGBUILDs for Arch and I'm going to update these to include all the patches and match the official builds. I'd like to prototype the above suggestions in a new branch and then we can decide if that works for us.

**Comment by qzed** [comment]
> What I do like about github releases though, is that it's hosted on Github infrastructure instead of a 3rd party host. We're all able to add new releases currently, but if @StollD becomes inactive or in the unlikely case that tmsp.io lapses, that's not good.

Right, we definitely should keep uploading releases on Github. That's not something I want to change. I'm just not convinced that maintaining a download script is worth it. But then again it shouldn't be that much work. Having the releases spread out and having users follow the links manually could have the benefit that the issues are opened on the corresponding repos and we wouldn't have to move it. Just feels a bit more explicit and we'd force the users to get at least a bit familiar with the different packages and maybe their goals.

> With Arch, you usually can only install the latest package version.

I thought about pushing the next 5.4 releases with a `-next` suffix. Just like the `-lts` one for 4.19. That way we can have all on the repo. I'm just not convinced that that particular suffix is the right choice. Maybe rather an `-ipts` for 5.3?

> I like the idea of a surface-firmware, but I'd like to include all the firmware we have (without linux-firmware). I don't think the user having to figure out what firmware packages they need is good UX, especially since it's just ~1-2MB per device.

Right, makes sense. I can live with that.

> I'm okay with a separate repo, although I find one monorepo a lot easier to manage.

Hmm yeah, there's that.

> I would like to have the build files that are used for the official packages available so that advanced users can reproduce the build without much extra hassle.

That would be nice. However, at the moment I'm using my Frankenstein script collection because that allows me to build the kernel directly from my checked-out tree (which makes development easier) and you definitely don't want to use that... I thought about restructuring that though. Ideally I'd like to build everything in a container (as I already need that for the .deb releases) from the upstream source + patches, but that adds some overhead. (A bit off-topic here, but I'd like to have some sort of local job scheduling/queuing system where I can just trigger new builds confined to containers and a fixed number of CPUs like with a CI system, so if you've got any tips there let me know. I just looked at docker and it already turns me off that I apparently need root access to run every container...)

> I'm already maintaining my own PKGBUILDs for Arch and I'm going to update these to include all the patches and match the official builds. I'd like to prototype the above suggestions in a new branch and then we can decide if that works for us.

Alright, go for it!

**Comment by PandorasFox** [comment]
> I just looked at docker and it already turns me off that I apparently need root access to run every container...

if your user is in the `docker` group, you shouldn't need root for container commands.

I think bazel or something would probably be decent for what you want since I think it can hide ~most of the docker stuff from you

**Comment by archseer** [comment]
> The docker group grants privileges equivalent to the root user. 

It's not recommended to add yourself to the docker group in most cases. They recently added a rootless mode, but it's experimental https://docs.docker.com/engine/security/rootless/

**Comment by qzed** [comment]
I'm currently thinking of putting the docker daemon inside an LXC container and do all that there... anyway that's for another time. Thanks for the hint to bazel, will definitely check that out.

**Comment by archseer** [comment]
Let me know what you think of https://github.com/linux-surface/linux-surface/tree/feature/pkg/pkg/arch

- [I cheated and concat'd the default `config` together with `surface.config`](https://github.com/linux-surface/linux-surface/blob/347670dc0aa51e75eb875167ed0828751b2e1acc/pkg/arch/kernel/PKGBUILD#L68-L70), maybe that can work without us having to run a merge script, that way I can keep the original files intact. 
- The config package should probably [run this](https://github.com/linux-surface/linux-surface/blob/3ca250a015f1096efc4e99d51ea382439bb30134/setup.sh#L48-L49) after installing.

With these, and your https://aur.archlinux.org/packages/libwacom-surface/ `setup.sh` becomes largely unnecessary. We can probably put these PKGBUILDs on AUR, for whoever wants to self-compile.

**Comment by StollD** [comment]
> However, at the moment I'm using my Frankenstein script collection because that allows me to build the kernel directly from my checked-out tree (which makes development easier) and you definitely don't want to use that...

Well, I was able to use that in the past just fine, and I don't even run Arch, lol.



> I just looked at docker and it already turns me off that I apparently need root access to run every container...

You could look at podman. I haven't used it myself extensively, and only through wrappers (fedora-toolbox), but it implements the same interface as docker, and is strictly rootless.

https://podman.io/

> [I cheated and concat'd the default `config` together with `surface.config`](https://github.com/linux-surface/linux-surface/blob/347670dc0aa51e75eb875167ed0828751b2e1acc/pkg/arch/kernel/PKGBUILD#L68-L70), maybe that can work without us having to run a merge script, that way I can keep the original files intact.

If you don't want to import the merge script from qzed, you should use `scripts/kconfig/merge_config.sh` to combine them, as that does some additional checking and already puts the config into the right place (`.config`).

Although it might be better to just link to the config from the config repo (or to the entire repository) in the PKGBUILD, and let `makepkg` download it for you, since those configs are already merged.

**Comment by qzed** [comment]
> If you don't want to import the merge script from qzed, you should use `scripts/kconfig/merge_config.sh` to combine them, as that does some additional checking and already puts the config into the right place (`.config`).

I think `scripts/kconfig/merge_config.sh` is the way to go for packages, especially if we've checked that combining the configs doesn't cause any issue or missing dependencies before. My script would add an extra dependency on `pipenv`.

**Comment by archseer** [comment]
Made the changes and updated the build to 5.4.8. My HID patch was backported so we should be able to drop it.

For the systemd unit, it seems that for Arch, it's recommended to have the user manually enable the units. So I added a post install script a notice.

**Comment by qzed** [comment]
> For the systemd unit, it seems that for Arch, it's recommended to have the user manually enable the units. So I added a post install script a notice.

Yeah, I think pretty much any config changes/automatic modification except for setting defaults for the software provided is discouraged. Which also makes the config package a bit complex if we want to ask the user for input, as that is a) discouraged and b) hard to implement (IIRC, I tried that a while back for the secure-boot-mok package).

Also do we want to provide PKGBUILDs for the other versions? That way we could auto-build all Arch packages from those and test them that way. In that case we need to handle kernel signing though... That's currently done via an extra patch but it might be better if we do that in the PKGBUILD instead. Or drop that from Arch packages and tell users to do that themselves, e.g to set up keys and use a pacman hook for signing.

**Comment by hasechris** [comment]
Hello, everybody,

@qzed Thank you very much for all this. The new organization github.com/linux-surface is in my opinion the best way.

I want to get involved in the project here too. Apart from the Surface Pro 6 I have an i7 8086K with 24GB of RAM that could be used with Gitlab CI for automatic builds during the night hours in Germany. The kernel builds will be relativly quick, it took me around 25 minutes to patch and build a manjaro flavoured surface kernel.

I've already looked at various options and think that automatic builds should not be a problem. I would create a repo in Gitlab (for my gitlab CI stages) and use the sources from the organization. Is there a way to upload the generated kernel packages back into the organization to a suitable location?

Regarding the installation of the (kernel) Packages: I think we should use the appropiate ways of the specific distro. e.g. for Arch the Arch User Repo should be used. for Debian Like Distros we should have a custom apt repo and so on. All would be filled from here on github. I dont think Bash-Scripts are a good way to distribute kernel packages and the required dependencies.

Regards
hasechris

**Comment by archseer** [comment]
Hi @hasechris, there's been some significant work done on this but via public infrastructure on Github Actions: https://github.com/linux-surface/linux-surface/tree/feature/pkg-repo

We plan to provide Arch, Debian and Fedora builds but probably we'd be open to some other distributions too if there's someone willing to step up as a maintainer for that distro.

- The bash script is already simplified down to a single hwclock tweak and will get removed completely. - - The firmware was split off and properly packaged
- We're working on an automatically updated repository and we'll push the PKGBUILDs to AUR as well of course

Once all these changes land it should be as easy as adding the repository and installing the necessary packages.
